feat: Add script functionality to transformer (#116)
grafana/killercoda: docs/transformer.md

1. A Killercoda maintainer reviews the PR to ensure that the generate tutorial works as expected.

> ## Adding foreground and background scripts (experimental)

Reviewer: ```suggestion
## Add foreground and background scripts
```

> > Note: It is recommended to use `foreground` scripts rather than `background` scripts. This allows the user to see the output of the script. 
> 
> Foreground and background scripts are shell scripts that run during the introduction and finish pages of a tutorial. The scripts are useful for setting up the environment for the tutorial and cleaning up after the tutorial. Scripts are another asset that needs to be maintained, so use them sparingly. A good example of using a script is to update the `docker compose` package before running the tutorial.

Reviewer: I think documenting the difference after introducing them is 
more valuable than a note about the difference before a reader knows what they are

```suggestion
Foreground and background scripts are shell scripts that run during the introduction and finish pages of a tutorial.
The scripts are useful for setting up the environment for the tutorial and cleaning up after the tutorial.

Scripts are another asset that needs to be maintained, so use them sparingly.
A good example of using a script is to update the Docker Compose package before running the tutorial.

Use foreground scripts when you want to the user to see the script output.
Use background scripts when you want to hide the output of the script.
```

### Create your script

> Since these scripts are primarily used for preparing the interactive environment, they are stored within the [`sandbox-scripts`](/sandbox-scripts/) directory in the Killercoda repository. Make sure to create your script in this directory and name with a unique name that reflects the purpose of the script. Note if a script already exists that performs the same function, you can reuse it.

Reviewer: Is this still correct?

> Since these scripts are primarily used for preparing the interactive environment, they are stored within the [`sandbox-scripts`](/sandbox-scripts/) directory in the Killercoda repository. Make sure to create your script in this directory and name with a unique name that reflects the purpose of the script. Note if a script already exists that performs the same function, you can reuse it.

Reviewer: I think this needs updating

Reviewer: ```suggestion
Create your script in the course directory sure with a name that reflects the purpose of the script.
```

> Here is an example of a script that updates the `docker-compose` package:

Reviewer: ```suggestion
Here is an example of a script that updates the Docker Compose package:
```

```bash
sudo install -m 0755 -d /etc/apt/keyrings
sudo curl -fsSL https://download.docker.com/linux/ubuntu/gpg -o /etc/apt/keyrings/docker.asc
sudo chmod a+r /etc/apt/keyrings/docker.asc
echo \
  "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.asc] https://download.docker.com/linux/ubuntu \
  $(. /etc/os-release && echo "$VERSION_CODENAME") stable" | \
  sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
sudo apt-get update
sudo apt-get install -y docker-compose-plugin && clear && echo "Setup complete. You may now begin the tutorial."

> ```

Reviewer: If we update the Tempo script, I think we should update this copy too.

> Note that you should add a message at the end of the script to inform the user that the setup is complete.

Reviewer: ```suggestion
> [!TIP]
> Add a message at the end of the script to inform the user that the setup is complete.
```

### Add the script to the tutorial

> To add the script to the tutorial, you need to add the script to the `killercoda` metadata in the source file. Note that we currently only support foreground and background scripts for the introduction and finish pages. These scripts are specified in the `killercoda.details` field in the source file. Lets look at an example:

Reviewer: ```suggestion
To add the script to the tutorial, you need to add the script to the `killercoda` metadata in the source file. 

> [!NOTE]
> The transformer tool only supports foreground and background scripts for the introduction and finish pages.

The following example sets the foreground script for the introduction page to be `docker-compose-update.sh`:
```

```yaml
title: Quick start for Tempo
menuTitle: Quick start for Tempo
description: Use Docker to quickly view traces using K-6 and Tempo
weight: 600
killercoda:
  title: Quick start for Tempo
  description: Use Docker to quickly view traces using K-6 and Tempo
  details:
      intro:
         foreground: docker-compose-update.sh
  backend:
    imageid: ubuntu
```

> In the example above, we have added a foreground script to the introduction page. The script is named `docker-compose-update.sh` and is located in the `sandbox-scripts` directory. The script will run when the introduction page is loaded.

Reviewer: ```suggestion
```

> Another example is adding a background script to the finish page:

Reviewer: ```suggestion
The following example sets the foreground script for the introduction page to be `docker-compose-cleanup.sh`:
```

```yaml
title: Quick start for Tempo
menuTitle: Quick start for Tempo
description: Use Docker to quickly view traces using K-6 and Tempo
weight: 600
killercoda:
  title: Quick start for Tempo
  description: Use Docker to quickly view traces using K-6 and Tempo
  details:
      finish:
         background: docker-compose-cleanup.sh
  backend:
    imageid: ubuntu
```

> 
> ### Quick FAQ
> 
> Here are a few common facts about using foreground and background scripts:
> 
> * Any scripts which are not specified in the `killercoda.details` field but exist in the course directory will be **deleted**. This helps us to clean up old scripts that are no longer in use.
> 
> * If you want to use the same script for multiple tutorials, you can specify the same script in multiple source files. The script will be copied to the respective tutorial directories.
> 
> * The transformer will fail if the script specified in the `killercoda.details` field does not exist in the `sandbox-scripts` directory.
> 
> * We currently only support foreground and background scripts for the introduction and finish pages. If you require scripts for other pages, please reach out to us via an issue.
> 

Reviewer: ```suggestion
```
